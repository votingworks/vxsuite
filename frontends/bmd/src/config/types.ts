import {
  BallotStyleId,
  CandidateContest,
  CandidateVote,
  ContestId,
  Contests,
  Election,
  ElectionDefinition,
  MachineId,
  MsEitherNeitherContest,
  OptionalVote,
  OptionalYesNoVote,
  PrecinctId,
  VotesDict,
  YesNoContest,
} from '@votingworks/types';
import { z } from 'zod';

// App
export const PrintOnly = {
  key: 'PrintOnly',
  productName: 'VxPrint',
  isMark: false,
  isPrint: true,
} as const;
export const MarkOnly = {
  key: 'MarkOnly',
  productName: 'VxMark',
  isMark: true,
  isPrint: false,
} as const;
export const MarkAndPrint = {
  key: 'MarkAndPrint',
  productName: 'VxMark',
  isPrint: true,
  isMark: true,
} as const;
export type AppMode = typeof MarkOnly | typeof PrintOnly | typeof MarkAndPrint;
export type AppModeKeys = AppMode['key'];
export const AppModeKeysSchema: z.ZodSchema<AppModeKeys> = z.union([
  z.literal(PrintOnly.key),
  z.literal(MarkOnly.key),
  z.literal(MarkAndPrint.key),
]);

export type ScreenOrientation = 'portrait' | 'landscape';

export interface MachineConfig {
  machineId: string;
  appMode: AppMode;
  codeVersion: string;
  screenOrientation: ScreenOrientation;
}

export interface MachineConfigResponse {
  machineId: string;
  appModeKey: AppModeKeys;
  codeVersion: string;
  screenOrientation: ScreenOrientation;
}
export const MachineConfigResponseSchema: z.ZodSchema<MachineConfigResponse> =
  z.object({
    machineId: MachineId,
    appModeKey: AppModeKeysSchema,
    codeVersion: z.string().nonempty(),
    screenOrientation: z.union([z.literal('portrait'), z.literal('landscape')]),
  });

export function getAppMode(key: AppModeKeys): AppMode {
  switch (key) {
    case PrintOnly.key:
      return PrintOnly;
    case MarkOnly.key:
      return MarkOnly;
    case MarkAndPrint.key:
      return MarkAndPrint;
    default:
      throw new Error(`unknown app mode: ${key}`);
  }
}

export type PostVotingInstructions = 'card' | 'cardless';

// Events
export type EventTargetFunction = (event: React.FormEvent<EventTarget>) => void;
export type InputChangeEventFunction =
  React.ChangeEventHandler<HTMLInputElement>;
export type TextareaChangeEventFunction =
  React.ChangeEventHandler<HTMLTextAreaElement>;
export type SelectChangeEventFunction =
  React.ChangeEventHandler<HTMLSelectElement>;

// Ballot
export type UpdateVoteFunction = (
  contestId: ContestId,
  vote: OptionalVote
) => void;
export type MarkVoterCardFunction = () => Promise<boolean>;
export interface BallotContextInterface {
  machineConfig: MachineConfig;
  ballotStyleId?: BallotStyleId;
  contests: Contests;
  readonly electionDefinition?: ElectionDefinition;
  isCardlessVoter: boolean;
  isLiveMode: boolean;
  markVoterCardPrinted: MarkVoterCardFunction;
  markVoterCardVoided: MarkVoterCardFunction;
  precinctId?: PrecinctId;
  resetBallot: (instructions?: PostVotingInstructions) => void;
  setUserSettings: SetUserSettings;
  updateTally: () => void;
  updateVote: UpdateVoteFunction;
  forceSaveVote: () => void;
  userSettings: UserSettings;
  votes: VotesDict;
}

// Review and Printed Ballot
export interface CandidateContestResultInterface {
  contest: CandidateContest;
  election: Election;
  precinctId: PrecinctId;
  vote: CandidateVote;
}
export interface YesNoContestResultInterface {
  contest: YesNoContest;
  vote: OptionalYesNoVote;
}
export interface MsEitherNeitherContestResultInterface {
  contest: MsEitherNeitherContest;
  eitherNeitherContestVote: OptionalYesNoVote;
  pickOneContestVote: OptionalYesNoVote;
}

export interface PrintOptions extends KioskBrowser.PrintOptions {
  sides: KioskBrowser.PrintSides;
}

// User Interface
export type ScrollDirections = 'up' | 'down';
export interface ScrollShadows {
  showBottomShadow: boolean;
  showTopShadow: boolean;
}
export interface Scrollable {
  isScrollable: boolean;
}

export type TextSizeSetting = 0 | 1 | 2 | 3;

export interface UserSettings {
  showSettingsModal: boolean;
  textSize: TextSizeSetting;
}
export type SetUserSettings = (partial: PartialUserSettings) => void;
export type PartialUserSettings = Partial<UserSettings>;

// Screen Reader
export interface SpeakOptions {
  now?: boolean;
}

export interface TextToSpeech {
  /**
   * Directly triggers speech of text. Resolves when speaking is done.
   */
  speak(text: string, options?: SpeakOptions): Promise<void>;

  /**
   * Stops any speaking that is currently happening.
   */
  stop(): void;

  /**
   * Prevents any sound from being made but otherwise functions normally.
   */
  mute(): void;

  /**
   * Allows sounds to be made.
   */
  unmute(): void;

  /**
   * Checks whether this TTS is muted.
   */
  isMuted(): boolean;

  /**
   * Toggles muted state, or sets it according to the argument.
   */
  toggleMuted(muted?: boolean): void;

  /**
   * Changes the current volume setting either up or down.
   */
  changeVolume?(): void;
}

/**
 * Implement this to provide screen reading.
 */
export interface ScreenReader {
  /**
   * Call this with an event target when a focus event occurs. Resolves when speaking is done.
   */
  onFocus(target?: EventTarget): Promise<void>;

  /**
   * Call this with an event target when a click event occurs. Resolves when speaking is done.
   */
  onClick(target?: EventTarget): Promise<void>;

  /**
   * Call this when a page load occurs. Resolves when speaking is done.
   */
  onPageLoad(): Promise<void>;

  /**
   * Enables the screen reader and announces the change. Resolves when speaking
   * is done.
   */
  enable(): Promise<void>;

  /**
   * Disables the screen reader and announces the change. Resolves when speaking
   * is done.
   */
  disable(): Promise<void>;

  /**
   * Toggles the screen reader being enabled and announces the change. Resolves
   * when speaking is done.
   */
  toggle(enabled?: boolean): Promise<void>;

  /**
   * Prevents any sound from being made but otherwise functions normally.
   */
  mute(): void;

  /**
   * Allows sounds to be made.
   */
  unmute(): void;

  /**
   * Checks whether this TTS is muted.
   */
  isMuted(): boolean;

  /**
   * Toggles muted state, or sets it according to the argument.
   */
  toggleMuted(muted?: boolean): void;

  /**
   * Directly triggers speech of text. Resolves when speaking is done.
   */
  speak(text: string, options?: SpeakOptions): Promise<void>;

  /**
   * Directly triggers speech of an element. Resolves when speaking is done.
   */
  speakNode(element: Element, options?: SpeakOptions): Promise<void>;

  /**
   * Directly triggers speech of an event target. Resolves when speaking is done.
   */
  speakEventTarget(target?: EventTarget, options?: SpeakOptions): Promise<void>;

  /**
   * Changes the current volume setting either up or down.
   */
  changeVolume(): void;
}

export interface VoiceSelector {
  (): SpeechSynthesisVoice | undefined;
}
