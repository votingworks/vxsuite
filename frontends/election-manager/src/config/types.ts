import { Admin } from '@votingworks/api';
import {
  BallotStyleId,
  CastVoteRecord,
  ContestTallyMeta,
  Dictionary,
  MachineId,
  Optional,
  PartyId,
  PrecinctId,
  PromiseOr,
  VotingMethod,
} from '@votingworks/types';
import { throwIllegalValue } from '@votingworks/utils';
import { z } from 'zod';

// Events
export type EventTargetFunction = (event: React.FormEvent<EventTarget>) => void;
export type InputEventFunction = (
  event: React.FormEvent<HTMLInputElement>
) => PromiseOr<void>;
export type TextareaEventFunction = (
  event: React.FormEvent<HTMLTextAreaElement>
) => PromiseOr<void>;
export type ButtonEventFunction = (
  event: React.MouseEvent<HTMLButtonElement>
) => PromiseOr<void>;

// Election
export type SaveElection = (electionJson: string) => Promise<void>;
export type ResetElection = () => Promise<void>;

export const PrintableBallotType = {
  Absentee: 'absentee',
  Precinct: 'standard',
} as const;
export type PrintableBallotType =
  typeof PrintableBallotType[keyof typeof PrintableBallotType];

export const PrintableBallotTypeSchema = z.union([
  z.literal('absentee'),
  z.literal('standard'),
]);

export function ballotModeToReadableString(
  ballotMode: Admin.BallotMode
): string {
  switch (ballotMode) {
    case Admin.BallotMode.Draft: {
      return 'Draft';
    }
    case Admin.BallotMode.Official: {
      return 'Official';
    }
    case Admin.BallotMode.Sample: {
      return 'Sample';
    }
    case Admin.BallotMode.Test: {
      return 'Test';
    }
    /* istanbul ignore next: Compile-time check for completeness */
    default: {
      throwIllegalValue(ballotMode);
    }
  }
}

export interface PrintOptions extends KioskBrowser.PrintOptions {
  sides: KioskBrowser.PrintSides;
}
export interface Printer {
  print(options: PrintOptions): Promise<void>;
}

// Router Props
export interface BallotScreenProps {
  ballotStyleId: BallotStyleId;
  precinctId: PrecinctId;
  localeCode?: string;
}
export interface PrecinctReportScreenProps {
  precinctId: PrecinctId;
}
export interface ScannerReportScreenProps {
  scannerId: string;
}
export interface BatchReportScreenProps {
  batchId: string;
}
export interface PartyReportScreenProps {
  partyId: PartyId;
}
export interface VotingMethodReportScreenProps {
  votingMethod: string;
}
export interface ManualDataPrecinctScreenProps {
  precinctId: PrecinctId;
}
export interface SmartcardsScreenProps {
  smartcardType: string;
}

// Tallies
export interface ExportableContestTally {
  readonly tallies: Dictionary<number>;
  readonly metadata: ContestTallyMeta;
}
export type ExportableTally = Dictionary<ExportableContestTally>;
export interface ExportableTallies {
  readonly talliesByPrecinct: Dictionary<ExportableTally>;
}

export interface ExternalFileConfiguration {
  readonly file: File;
  readonly votingMethod: VotingMethod;
}

export enum ResultsFileType {
  // eslint-disable-next-line @typescript-eslint/no-shadow
  CastVoteRecord = 'cvr',
  SEMS = 'sems',
  All = 'all',
  Manual = 'manual',
}
export type OptionalFile = Optional<File>;

export interface CastVoteRecordFile {
  readonly name: string;
  readonly importedCvrCount: number;
  readonly duplicatedCvrCount: number;
  readonly scannerIds: readonly string[];
  readonly precinctIds: readonly PrecinctId[];
  readonly allCastVoteRecords: readonly CastVoteRecord[];
  readonly exportTimestamp: Date;
}
export interface CastVoteRecordFilePreprocessedData {
  readonly name: string;
  readonly path: string;
  readonly cvrCount: number;
  readonly scannerIds: readonly string[];
  readonly exportTimestamp: Date;
  readonly isTestModeResults: boolean;
}

export type VoteCounts = Dictionary<Dictionary<number>>;
export type OptionalVoteCounts = Optional<Dictionary<Dictionary<number>>>;

export type Iso8601Timestamp = string;

export type ConverterClientType = 'ms-sems' | 'nh-accuvote';

export const ConverterClientTypeSchema = z.union([
  z.literal('ms-sems'),
  z.literal('nh-accuvote'),
]);

export type SmartcardType = 'election' | 'system-administrator';
export const SmartcardTypeRegExPattern = '(election|system-administrator)';

export interface MachineConfig {
  machineId: string;
  codeVersion: string;
}

export const MachineConfigSchema: z.ZodSchema<MachineConfig> = z.object({
  machineId: MachineId,
  codeVersion: z.string().nonempty(),
});
