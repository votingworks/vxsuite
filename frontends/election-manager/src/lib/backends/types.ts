import {
  ElectionDefinition,
  ExternalTallySourceType,
  FullElectionExternalTallies,
  FullElectionExternalTally,
  Iso8601Timestamp,
} from '@votingworks/types';
import { PrintedBallot } from '../../config/types';
import { CastVoteRecordFiles } from '../../utils/cast_vote_record_files';

export interface AddCastVoteRecordFileResult {
  readonly wasExistingFile: boolean;
  readonly newlyAdded: number;
  readonly alreadyPresent: number;
}

export interface ElectionManagerStoreBackend {
  /**
   * Resets all stored data, including the election definition and CVRs.
   */
  reset(): Promise<void>;

  /**
   * Loads the existing election definition, if there is one.
   */
  loadElectionDefinitionAndConfiguredAt(): Promise<
    | { electionDefinition: ElectionDefinition; configuredAt: Iso8601Timestamp }
    | undefined
  >;

  /**
   * Configures with a new election definition after resetting.
   *
   * @param newElectionData election definition as JSON string
   */
  configure(newElectionData: string): Promise<ElectionDefinition>;

  /**
   * Loads the existing cast vote record files.
   */
  loadCastVoteRecordFiles(): Promise<CastVoteRecordFiles | undefined>;

  /**
   * Adds a new cast vote record file.
   */
  addCastVoteRecordFile(
    newCastVoteRecordFile: File
  ): Promise<AddCastVoteRecordFileResult>;

  /**
   * Resets all cast vote record files.
   */
  clearCastVoteRecordFiles(): Promise<void>;

  /**
   * Loads the existing external tallies.
   */
  loadFullElectionExternalTallies(): Promise<
    FullElectionExternalTallies | undefined
  >;

  /**
   * Updates the external tally for a given source.
   */
  updateFullElectionExternalTally(
    sourceType: ExternalTallySourceType,
    newFullElectionExternalTally: FullElectionExternalTally
  ): Promise<void>;

  /**
   * Removes the external tally for a given source.
   */
  removeFullElectionExternalTally(
    sourceType: ExternalTallySourceType
  ): Promise<void>;

  /**
   * Clears all external tallies.
   */
  clearFullElectionExternalTallies(): Promise<void>;

  /**
   * Loads the existing setting for whether the results are official.
   */
  loadIsOfficialResults(): Promise<boolean | undefined>;

  /**
   * Marks the results as official. No more tallies can be added after this.
   */
  markResultsOfficial(): Promise<void>;

  /**
   * Loads the existing printed ballots.
   */
  loadPrintedBallots(): Promise<PrintedBallot[] | undefined>;

  /**
   * Adds a new printed ballot to the list.
   */
  addPrintedBallot(printedBallot: PrintedBallot): Promise<void>;
}
