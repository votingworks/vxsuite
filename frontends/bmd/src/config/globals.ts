export const IDLE_TIMEOUT_SECONDS = 5 * 60; // VVSG Requirement: 2–5 minutes
export const IDLE_RESET_TIMEOUT_SECONDS = 45; // VVSG Requirement: 20–45 seconds
export const RECENT_PRINT_EXPIRATION_SECONDS = 1 * 60; // 1 minute
export const CARD_LONG_VALUE_WRITE_DELAY = 1000;
export const BALLOT_PRINTING_TIMEOUT_SECONDS = 5;
export const BALLOT_INSTRUCTIONS_TIMEOUT_SECONDS = 30;
export const REPORT_PRINTING_TIMEOUT_SECONDS = 4;
export const CHECK_ICON = '✓';
export const FONT_SIZES = [22, 28, 36, 48];
export const DEFAULT_FONT_SIZE = 1;
export const LARGE_DISPLAY_FONT_SIZE = 3;
export const TEXT_SIZE = 1;
export const WRITE_IN_CANDIDATE_MAX_LENGTH = 40;
export const QUIT_KIOSK_IDLE_SECONDS = 5 * 60; // 5 minutes
export const SECURITY_PIN_LENGTH = 6;
