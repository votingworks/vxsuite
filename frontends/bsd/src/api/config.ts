import {
  ElectionDefinition,
  MarkThresholds,
  Optional,
  Precinct,
  safeParseJson,
  unsafeParse,
} from '@votingworks/types';
import { ErrorsResponse, OkResponse, Scan } from '@votingworks/api';
import { fetchJson } from '@votingworks/utils';

async function patch<Body extends string | ArrayBuffer | unknown>(
  url: string,
  value: Body
): Promise<void> {
  const isJson =
    typeof value !== 'string' &&
    !(value instanceof ArrayBuffer) &&
    !(value instanceof Uint8Array);
  const response = await fetch(url, {
    method: 'PATCH',
    body: isJson ? JSON.stringify(value) : (value as BodyInit),
    headers: {
      'Content-Type': isJson ? 'application/json' : 'application/octet-stream',
    },
  });
  const body: OkResponse | ErrorsResponse = await response.json();

  if (body.status !== 'ok') {
    throw new Error(`PATCH ${url} failed: ${JSON.stringify(body.errors)}`);
  }
}

async function put<Body extends string | ArrayBuffer | unknown>(
  url: string,
  value: Body
): Promise<void> {
  const isJson =
    typeof value !== 'string' &&
    !(value instanceof ArrayBuffer) &&
    !(value instanceof Uint8Array);
  const response = await fetch(url, {
    method: 'PUT',
    body: isJson ? JSON.stringify(value) : (value as BodyInit),
    headers: {
      'Content-Type': isJson ? 'application/json' : 'application/octet-stream',
    },
  });
  const body: OkResponse | ErrorsResponse = await response.json();

  if (body.status !== 'ok') {
    throw new Error(`PUT ${url} failed: ${JSON.stringify(body.errors)}`);
  }
}

async function del(url: string): Promise<void> {
  const response = await fetch(url, {
    method: 'DELETE',
    headers: { 'Content-Type': 'application/json' },
  });
  const body: OkResponse | ErrorsResponse = await response.json();

  if (body.status !== 'ok') {
    throw new Error(`DELETE ${url} failed: ${JSON.stringify(body.errors)}`);
  }
}

export async function getElection(): Promise<string | undefined> {
  const response = await fetch('/central-scanner/config/election');

  if (response.status === 404) {
    return undefined;
  }

  return response.text();
}

export async function getElectionDefinition(): Promise<
  ElectionDefinition | undefined
> {
  return (
    (safeParseJson(
      await (
        await fetch('/central-scanner/config/election', {
          headers: { Accept: 'application/json' },
        })
      ).text(),
      Scan.GetElectionConfigResponseSchema
    ).unsafeUnwrap() as Exclude<Scan.GetElectionConfigResponse, string>) ??
    undefined
  );
}

export async function setElection(
  electionData?: string,
  { ignoreBackupRequirement }: { ignoreBackupRequirement?: boolean } = {}
): Promise<void> {
  if (typeof electionData === 'undefined') {
    let deletionUrl = '/central-scanner/config/election';
    if (ignoreBackupRequirement) {
      deletionUrl += '?ignoreBackupRequirement=true';
    }
    await del(deletionUrl);
  } else {
    await patch<Scan.PatchElectionConfigRequest>(
      '/central-scanner/config/election',
      new TextEncoder().encode(electionData)
    );
  }
}

export async function getTestMode(): Promise<boolean> {
  return safeParseJson(
    await (await fetch('/central-scanner/config/testMode')).text(),
    Scan.GetTestModeConfigResponseSchema
  ).unsafeUnwrap().testMode;
}

export async function setTestMode(testMode: boolean): Promise<void> {
  await patch<Scan.PatchTestModeConfigRequest>(
    '/central-scanner/config/testMode',
    {
      testMode,
    }
  );
  const newTestMode = await getTestMode();
  if (newTestMode !== testMode) {
    throw new Error('Error setting test mode, please try again');
  }
}

export async function getMarkThresholdOverrides(): Promise<
  MarkThresholds | undefined
> {
  const { markThresholdOverrides } = unsafeParse(
    Scan.GetMarkThresholdOverridesConfigResponseSchema,
    await fetchJson('/central-scanner/config/markThresholdOverrides')
  );
  return markThresholdOverrides;
}

export async function setMarkThresholdOverrides(
  markThresholdOverrides?: MarkThresholds
): Promise<void> {
  if (typeof markThresholdOverrides === 'undefined') {
    await del('/central-scanner/config/markThresholdOverrides');
  } else {
    await patch<Scan.PatchMarkThresholdOverridesConfigRequest>(
      '/central-scanner/config/markThresholdOverrides',
      { markThresholdOverrides }
    );
  }
}

export async function getCurrentPrecinctId(): Promise<
  Optional<Precinct['id']>
> {
  return safeParseJson(
    await (
      await fetch('/central-scanner/config/precinct', {
        headers: { Accept: 'application/json' },
      })
    ).text(),
    Scan.GetCurrentPrecinctResponseSchema
  ).unsafeUnwrap().precinctId;
}

export async function setCurrentPrecinctId(
  precinctId: Precinct['id']
): Promise<void> {
  await put<Scan.PutCurrentPrecinctConfigRequest>(
    '/central-scanner/config/precinct',
    {
      precinctId,
    }
  );
}
