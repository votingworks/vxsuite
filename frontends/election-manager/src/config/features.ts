import { unsafeParse } from '@votingworks/types';
import { ConverterClientType, ConverterClientTypeSchema } from './types';

/**
 * Determines which converter client to use, if any.
 */
export function getConverterClientType(): ConverterClientType | undefined {
  const rawConverterClientType = process.env.REACT_APP_VX_CONVERTER;

  if (!rawConverterClientType) {
    return;
  }

  return unsafeParse(ConverterClientTypeSchema, rawConverterClientType);
}
