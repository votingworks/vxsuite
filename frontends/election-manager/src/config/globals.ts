import { Dictionary } from '@votingworks/types';

export const CHECK_ICON = 'âœ“';
export const DEFAULT_LOCALE = 'en-US';
export const LANGUAGES: Dictionary<string> = {
  en: 'English',
  es: 'Spanish',
};
export const ABSENTEE_TINT_COLOR = 'rgb(180, 100, 225)';

export const TIME_FORMAT = 'MM/DD/YYYY hh:mm:ss A';
