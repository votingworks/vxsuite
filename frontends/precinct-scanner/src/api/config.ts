import {
  ElectionDefinition,
  Optional,
  MarkThresholds,
  safeParseJson,
  PrecinctSelection,
  PollsState,
} from '@votingworks/types';
import { ErrorsResponse, OkResponse, Scan } from '@votingworks/api';
import { BallotPackage, BallotPackageEntry, assert } from '@votingworks/utils';
import { EventEmitter } from 'events';

async function patch<Body extends string | ArrayBuffer | unknown>(
  url: string,
  value: Body
): Promise<void> {
  const isJson =
    typeof value !== 'string' &&
    !(value instanceof ArrayBuffer) &&
    !(value instanceof Uint8Array);
  const response = await fetch(url, {
    method: 'PATCH',
    body: isJson ? JSON.stringify(value) : (value as BodyInit),
    headers: {
      'Content-Type': /* istanbul ignore next */ isJson
        ? 'application/json'
        : 'application/octet-stream',
    },
  });
  const body: OkResponse | ErrorsResponse = await response.json();

  if (body.status !== 'ok') {
    throw new Error(`PATCH ${url} failed: ${JSON.stringify(body.errors)}`);
  }
}

async function put<Body extends string | ArrayBuffer | unknown>(
  url: string,
  value: Body
): Promise<void> {
  const isJson =
    typeof value !== 'string' &&
    !(value instanceof ArrayBuffer) &&
    !(value instanceof Uint8Array);
  const response = await fetch(url, {
    method: 'PUT',
    body: /* istanbul ignore next */ isJson
      ? JSON.stringify(value)
      : (value as BodyInit),
    headers: {
      'Content-Type': isJson ? 'application/json' : 'application/octet-stream',
    },
  });
  const body: OkResponse | ErrorsResponse = await response.json();

  if (body.status !== 'ok') {
    throw new Error(`PUT ${url} failed: ${JSON.stringify(body.errors)}`);
  }
}

async function del(url: string): Promise<void> {
  const response = await fetch(url, {
    method: 'DELETE',
    headers: { 'Content-Type': 'application/json' },
  });
  const body: OkResponse | ErrorsResponse = await response.json();

  if (body.status !== 'ok') {
    throw new Error(`DELETE ${url} failed: ${JSON.stringify(body.errors)}`);
  }
}

export async function getElectionDefinition(): Promise<
  ElectionDefinition | undefined
> {
  return (
    (safeParseJson(
      await (
        await fetch('/precinct-scanner/config/election', {
          headers: { Accept: 'application/json' },
        })
      ).text(),
      Scan.GetElectionConfigResponseSchema
    ).unsafeUnwrap() as Exclude<Scan.GetElectionConfigResponse, string>) ??
    undefined
  );
}

export async function setElection(
  electionData?: string,
  { ignoreBackupRequirement }: { ignoreBackupRequirement?: boolean } = {}
): Promise<void> {
  if (typeof electionData === 'undefined') {
    let deletionUrl = '/precinct-scanner/config/election';
    if (ignoreBackupRequirement) {
      deletionUrl += '?ignoreBackupRequirement=true';
    }
    await del(deletionUrl);
  } else {
    // TODO(528) add proper typing here
    await patch('/precinct-scanner/config/election', electionData);
  }
}

export async function getTestMode(): Promise<boolean> {
  return safeParseJson(
    await (await fetch('/precinct-scanner/config/testMode')).text(),
    Scan.GetTestModeConfigResponseSchema
  ).unsafeUnwrap().testMode;
}

export async function setTestMode(testMode: boolean): Promise<void> {
  await patch<Scan.PatchTestModeConfigRequest>(
    '/precinct-scanner/config/testMode',
    {
      testMode,
    }
  );
}

export async function getMarkThresholds(): Promise<Optional<MarkThresholds>> {
  return safeParseJson(
    await (
      await fetch('/precinct-scanner/config/markThresholdOverrides', {
        headers: { Accept: 'application/json' },
      })
    ).text(),
    Scan.GetMarkThresholdOverridesConfigResponseSchema
  ).unsafeUnwrap().markThresholdOverrides;
}

export async function setMarkThresholdOverrides(
  markThresholdOverrides?: MarkThresholds
): Promise<void> {
  if (typeof markThresholdOverrides === 'undefined') {
    await del('/precinct-scanner/config/markThresholdOverrides');
  } else {
    await patch<Scan.PatchMarkThresholdOverridesConfigRequest>(
      '/precinct-scanner/config/markThresholdOverrides',
      { markThresholdOverrides }
    );
  }
}

export async function getPrecinctSelection(): Promise<
  Optional<PrecinctSelection>
> {
  return safeParseJson(
    await (
      await fetch('/precinct-scanner/config/precinct', {
        headers: { Accept: 'application/json' },
      })
    ).text(),
    Scan.GetPrecinctSelectionConfigResponseSchema
  ).unsafeUnwrap().precinctSelection;
}

export async function setPrecinctSelection(
  precinctSelection: PrecinctSelection
): Promise<void> {
  await put<Scan.PutPrecinctSelectionConfigRequest>(
    '/precinct-scanner/config/precinct',
    {
      precinctSelection,
    }
  );
}

export async function getPollsState(): Promise<PollsState> {
  return safeParseJson(
    await (
      await fetch('/precinct-scanner/config/polls', {
        headers: { Accept: 'application/json' },
      })
    ).text(),
    Scan.GetPollsStateConfigResponseSchema
  ).unsafeUnwrap().pollsState;
}

export async function setPollsState(pollsState: PollsState): Promise<void> {
  await put<Scan.PutPollsStateConfigRequest>('/precinct-scanner/config/polls', {
    pollsState,
  });
}

export interface AddTemplatesEvents extends EventEmitter {
  on(
    event: 'configuring',
    callback: (
      pkg: BallotPackage,
      electionDefinition: ElectionDefinition
    ) => void
  ): this;
  on(
    event: 'uploading',
    callback: (pkg: BallotPackage, entry: BallotPackageEntry) => void
  ): this;
  on(event: 'completed', callback: (pkg: BallotPackage) => void): this;
  on(event: 'error', callback: (error: Error) => void): this;
  off(
    event: 'configuring',
    callback: (
      pkg: BallotPackage,
      electionDefinition: ElectionDefinition
    ) => void
  ): this;
  off(
    event: 'uploading',
    callback: (pkg: BallotPackage, entry: BallotPackageEntry) => void
  ): this;
  off(event: 'completed', callback: (pkg: BallotPackage) => void): this;
  off(event: 'error', callback: (error: Error) => void): this;
  emit(
    event: 'configuring',
    pkg: BallotPackage,
    electionDefinition: ElectionDefinition
  ): boolean;
  emit(
    event: 'uploading',
    pkg: BallotPackage,
    entry: BallotPackageEntry
  ): boolean;
  emit(event: 'completed', pkg: BallotPackage): boolean;
  emit(event: 'error', error: Error): boolean;
}

export function addTemplates(pkg: BallotPackage): AddTemplatesEvents {
  const result: AddTemplatesEvents = new EventEmitter();

  setImmediate(async () => {
    try {
      result.emit('configuring', pkg, pkg.electionDefinition);
      await setElection(pkg.electionDefinition.electionData);

      for (const ballot of pkg.ballots) {
        result.emit('uploading', pkg, ballot);

        const body = new FormData();

        body.append(
          'ballots',
          new Blob([ballot.pdf], { type: 'application/pdf' }),
          ballot.ballotConfig.filename
        );

        body.append(
          'metadatas',
          new Blob([JSON.stringify(ballot.ballotConfig)], {
            type: 'application/json',
          }),
          'ballot-config.json'
        );

        if (ballot.layout) {
          body.append(
            'layouts',
            new Blob([JSON.stringify(ballot.layout)], {
              type: 'application/json',
            }),
            ballot.ballotConfig.layoutFilename
          );
        }

        await fetch('/precinct-scanner/config/addTemplates', {
          method: 'POST',
          body,
        });
      }

      result.emit('completed', pkg);
    } catch (error) {
      assert(error instanceof Error);
      result.emit('error', error);
    }
  });

  return result;
}
