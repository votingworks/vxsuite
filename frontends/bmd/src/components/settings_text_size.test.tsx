import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import * as GLOBALS from '../config/globals';
import { UserSettings } from '../config/types';
import { SettingsTextSize } from './settings_text_size';

const userSettings: UserSettings = GLOBALS.DEFAULT_USER_SETTINGS;
const setUserSettings = jest.fn();

it('renders SettingsTextSize', () => {
  const { container, getAllByText } = render(
    <SettingsTextSize
      userSettings={userSettings}
      setUserSettings={setUserSettings}
    />
  );
  expect(container.firstChild).toMatchSnapshot();
  const buttons = getAllByText('A');
  expect(buttons.length).toBe(3);
  fireEvent.click(buttons[0]);
  expect(setUserSettings).toBeCalledWith({ textSize: 0 });
  fireEvent.click(buttons[1]);
  expect(setUserSettings).toBeCalledWith({ textSize: 1 });
  fireEvent.click(buttons[2]);
  expect(setUserSettings).toBeCalledWith({ textSize: 2 });
});
