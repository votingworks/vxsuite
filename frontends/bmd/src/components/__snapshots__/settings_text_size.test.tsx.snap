// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SettingsTextSize 1`] = `
.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  white-space: nowrap;
}

.c1 > button {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  box-shadow: inset 1px 0 0 rgb(190,190,190);
}

.c1 > button:first-child {
  box-shadow: none;
}

.c1 > button:not(:last-child) {
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

.c1 > button:not(:first-child) {
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}

.c1 > button:disabled {
  background: #028099;
  color: #ffffff;
}

.c1 button {
  min-width: 98px;
}

.c1 button[data-size='0'] {
  font-size: 22px !important;
}

.c1 button[data-size='1'] {
  font-size: 28px !important;
}

.c1 button[data-size='2'] {
  font-size: 36px !important;
}

.c0 {
  display: block;
  margin-bottom: 0.4rem;
}

<p>
  <label
    aria-hidden="true"
    class="c0"
  >
    Text Size
  </label>
  <span
    class="sc-bqiQRQ c1"
    data-testid="change-text-size-buttons"
  >
    <button
      aria-label=" Text Size: Small"
      class="sc-egiSv feoAXt"
      data-size="0"
      type="button"
      value="0"
    >
      A
    </button>
    <button
      aria-label="Selected Text Size: Medium"
      class="sc-egiSv dGzrLr"
      data-size="1"
      type="button"
      value="1"
    >
      A
    </button>
    <button
      aria-label=" Text Size: Large"
      class="sc-egiSv feoAXt"
      data-size="2"
      type="button"
      value="2"
    >
      A
    </button>
  </span>
</p>
`;
